xss-lock: improve locker options passing

turboMaCk opened the pull request:
*Noob alert: This is my first PR to nixpkgs.*

###### Motivation for this change
Currently, command line arguments can't be passed in `locker` easily without workarounds.

```nix
# Doesn't work
programs.xss-lock.lockerCommand = "${pkgs.xlockmore}/bin/xlock -mode ant";
```

This is because xss-lock doesn't recognize these are locker arguments.

```nix
# Works but seems hacky
programs.xss-lock.lockerCommand = "-- ${pkgs.xlockmore}/bin/xlock -mode ant";
```

###### Things done

In short:

- allow locker options without hacks
- add extraOptions

Locker is now separated by `--`. Because of this one can no longer abuse `locker` option to pass arguments to `xss-lock` and therefore new option `extraOptions` was added to pass options to `xss-lock`

**This is a breaking change for configurations which used workaround described**

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

turboMaCk committed:
xss-locker: improve options passing

- allow locker options without hacks
- add extraOptions

turboMaCk commented:
cc @ma27 who is listed maintainer of the `xss-lock` package.

Ma27 commented on nixos/modules/programs/xss-lock.nix:
```diff
@@ -8,20 +8,37 @@ in
+            "${pkgs.xss-lock}/bin/xss-lock"
+          ] ++ cfg.extraOptions ++ [
+            "--"
+            "cfg.lockerCommand"
```
I guess you didn't mean to quote this :)

turboMaCk replied:
wow of course. I wonder how is it possible it worked for me after rebuild :thinking:

turboMaCk replied:
ok now I see. I've restarted `xauto-lock` service instead of `xss-lock` one :man_facepalming: Going to do some more testing.

Ma27 commented:
@GrahamcOfBorg test xss-lock

Ma27 commented:
Bt win case another maintainer sees this: I'd like to add a simple testcase for the example provided above before merging this, but I'm slightly busy with other stuff now, so it would be great if you could be patient until I've pushed that :)

turboMaCk commented:
I really appreciate how responsive (and fast) you were. There's no need to rush this imo. Better right than right now.

Also, let me know if I should change or do something extra here. I'm interested in learning how to work with nixpkgs so I can contribute more in the future.

turboMaCk commented on nixos/modules/programs/xss-lock.nix:
```diff
@@ -8,20 +8,37 @@ in
       description = "Locker to be used with xsslock";
     };
+
+    extraOptions = mkOption {
```
this is named `extraOptions` but there are no other options. I did this because of consistency with `xautolock` but it might make more sense to name it `options`. Some with more experience should decide.

Ma27 replied:
`extraOptions` is usually used when adding some special service options in the verbatim format for the software itself (in this case CLI args). The default and most important option is `lockerCommand`, so it's perfectly fine to specify all the remaining options using `extraOptions`. :+1:

Ma27 committed:
nixos/xss-lock: improve module

* Don't use `literalExample`, raw Nix values can directly be specified
  as an option example which provides support for highlighting in the
  manual as well.

* Escape shell args for `extraOptions`: I.e. the `-n` option might be
  problematic as a longer notification command might be misinterpreted.

Ma27 committed:
nixos/xss-lock: add testcase for `lockerCommand` with several CLI options.

Ma27 commented:
@GrahamcOfBorg test xss-lock

Ma27 commented:
@turboMaCk thanks!
